btree/btree.c

```diff
@@ -4466,12 +4492,425 @@ static void m0_btree_ut_multi_stream_kv_oper(void)
+	m0_be_tx_prep(tx, NULL);
+
+	key_iter = ti->ti_key_first;
+
+	rec.r_key.k_data   = M0_BUFVEC_INIT_BUF(&k_ptr, &ksize);
+	rec.r_val          = M0_BUFVEC_INIT_BUF(&v_ptr, &vsize);
+
+	data.key           = &rec.r_key;
+	data.value         = &rec.r_val;
+
+	ut_cb.c_act        = btree_kv_put_cb;
+	ut_cb.c_datum      = &data;
+
+	get_rec.r_key.k_data   = M0_BUFVEC_INIT_BUF(&get_k_ptr, &get_ksize);
+	get_rec.r_val          = M0_BUFVEC_INIT_BUF(&get_v_ptr, &get_vsize);
+
+	get_data.key           = &get_rec.r_key;
+	get_data.value         = &get_rec.r_val;
+	get_data.check_value   = true;
+
+	ut_get_cb.c_act        = btree_kv_get_cb;
+	ut_get_cb.c_datum      = &get_data;
+
+	tree                   = ti->ti_tree;
+
+	/** Wait till all the threads have been initialised. */
+	while (!thread_start)
+	;
+
+	while (key_iter <= ti->ti_key_last) {
+		uint64_t  key_start;
+		uint64_t  key_end;
+		uint64_t  keys_put_count = 0;
+		uint64_t  keys_found_count = 0;
+		int       i;
+		int32_t   r;
+		uint64_t  iter_dir;
+		uint64_t  del_key;
+
+		key_start = key_iter;
+		if (ti->ti_random_bursts) {
+			random_r(&ti->ti_random_buf, &r);
+			key_end = (r % (ti->ti_key_last - key_start)) +
+				  key_start;
+		} else
+			key_end = ti->ti_key_last;
+
+		/** PUT keys and their corresponding values in the tree. */
+		while (key_start <= key_end) {
+			/**
+			 *  Embed the thread-id in LSB so that different threads
+			 *  will target the same node thus causing race
+			 *  conditions useful to mimic and test btree operations
+			 *  in a loaded system.
+			 */
+			key[0] = (key_start << (sizeof(ti->ti_thread_id) * 8)) +
+				ti->ti_thread_id;
+			key[0] = m0_byteorder_cpu_to_be64(key[0]);
+			for (i = 1; i < ARRAY_SIZE(key); i++)
+				key[i] = key[0];
+
+			value[0] = key[0];
+			for (i = 1; i < ARRAY_SIZE(value); i++)
+				value[i] = value[0];
+
+			M0_BTREE_OP_SYNC_WITH_RC(&kv_op.bo_op,
+						 m0_btree_put(tree, tx,
+							      &rec, &ut_cb,
+							      0, &kv_op),
+						 &kv_op.bo_sm_group,
+						 &kv_op.bo_op_exec);
+			M0_ASSERT(data.flags == 0);
+
+			keys_put_count++;
+			key_start++;
+		}
+
+		/** GET and ITERATE over the keys which we inserted above. */
+
+		/**  Randomly decide the iteration direction. */
+		random_r(&ti->ti_random_buf, &r);
+
+		key_start = key_iter;
+		if (r%2) {
+			/** Iterate forward. */
+			iter_dir = BOF_NEXT;
+			key[0] = (key_start <<
+				      (sizeof(ti->ti_thread_id) * 8)) +
+				     ti->ti_thread_id;
+			key[0] = m0_byteorder_cpu_to_be64(key[0]);
+			for (i = 1; i < ARRAY_SIZE(key); i++)
+				key[i] = key[0];
+		} else {
+			/** Iterate backward. */
+			iter_dir = BOF_PREV;
+			key[0] = (key_end <<
+				      (sizeof(ti->ti_thread_id) * 8)) +
+				     ti->ti_thread_id;
+			key[0] = m0_byteorder_cpu_to_be64(key[0]);
+			for (i = 1; i < ARRAY_SIZE(key); i++)
+				key[i] = key[0];
+		}
+
+		M0_BTREE_OP_SYNC_WITH_RC(&kv_op.bo_op,
+					 m0_btree_get(tree,
+						      &rec.r_key, &ut_get_cb,
+						      iter_dir, &kv_op),
```

Comment:
for `m0_btree_get` flag should be either `BOF_EQUAL` or `BOF_SLANT`.

Reply:
Good catch.